### PR TITLE
Assignment7+8 Manny Aguilera & Paulson Ray Thompson

### PR DESCRIPTION
**Pull request by Purplepigs0129**
Manny's Git - @MannyPoptart
Paulson's Git - @Purplepigs0129

### CODE REVIEW

**Comment by Tfelchlin13** [comment]
Ill review this one

File: Assignment.Tests/SampleDataTests.cs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+    [TestMethod]
+    public void GetAggregateSortedListOfStatesUsingCsvRows_SortedAndUnique_ReturnsSuccessful()
+    {
+        string uniqueSortedStates = string.Join(", ", SampleData.GetAggregateSortedListOfStatesUsingCsvRows());
+        Assert.AreEqual("AL, AZ, CA, DC, FL, GA, IN, KS, LA, MD, MN, MO, MT, NC, NE, NH, NV, NY, OR, PA, SC, TN, TX, UT, VA, WA, WV", uniqueSortedStates);
+    }
+
+    [TestMethod]
+    public void People_CountIs50_ReturnsSuccessful()
```
Comment (Tfelchlin13) [inline Assignment.Tests/SampleDataTests.cs:41]: looks good

File: Directory.Build.props [hunk 2]
```diff
@@ -1,9 +1,10 @@
...
-    <TargetFramework>net6.0</TargetFramework>
-    <Nullable>enable</Nullable>
-    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
-    <EnableNETAnalyzers>true</EnableNETAnalyzers>
-    <CodeAnalysisTreatWarningsAsErrors>true</CodeAnalysisTreatWarningsAsErrors>
-  </PropertyGroup>
+<Project>
+<PropertyGroup>
+    <TargetFramework>net8.0</TargetFramework>
+	<LangVersion>12.0</LangVersion>
```
Comment (Tfelchlin13) [inline Directory.Build.props:4]:
```suggestion
```

File: Directory.Build.props [hunk 3]
```diff
@@ -1,9 +1,10 @@
...
-<PropertyGroup>
-    <TargetFramework>net6.0</TargetFramework>
-    <Nullable>enable</Nullable>
-    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
-    <EnableNETAnalyzers>true</EnableNETAnalyzers>
-    <CodeAnalysisTreatWarningsAsErrors>true</CodeAnalysisTreatWarningsAsErrors>
-  </PropertyGroup>
+<Project>
+<PropertyGroup>
+    <TargetFramework>net8.0</TargetFramework>
```
Comment (Tfelchlin13) [inline Directory.Build.props:3]:
```suggestion
    <TargetFramework>net8.0</TargetFramework>
    <LangVersion>12.0</LangVersion>
```
fixed spacing

**Review by Tfelchlin13 — COMMENTED** [review]
Honestly your code looked good and you have great code coverage for your tests overall i loved how you guys code this. 👍

Implement the ISampleData.CsvRows property, loading the data from the People.csv file and returning each line as a single string. ✔

Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔
Using LINQ, skip the first row in the People.csv. ✔
Be sure to appropriately handle resource (IDisposable) items correctly if applicable (and it may not be depending on how you implement it). ✔
Implement IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() to return a sorted, unique list of states. ✔

Use ISampleData.CsvRows for your data source. ✔
Don't forget the list should be unique. ✔
Sort the list alphabetically. ✔
Include a test that leverages a hardcoded list of addresses. ✔
Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ✔
Implement ISampleData.GetAggregateSortedListOfStatesUsingCsvRows() to return a string that contains a unique, comma separated list of states. ✔

Use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows() for your data source. ✔
Consider "selecting" only the states and calling ToArray() to retrieve an array of all the state names. ✔
Given the array, consider using string.Join to combine the list into a single string. ❌✔
Implement the ISampleData.People property to return all the items in People.csv as Person objects ✔

Use ISampleData.CsvRows as the source of the data. ✔
Sort the list by State, City, and Zip. (Sort the addresses first then select). ✔
Be sure that Person.Address is also populated. ✔
Adding null validation to all the Person and Address properties is optional.
Consider using ISampleData.CsvRows in your test to verify your results. ✔
Implement ISampleDate.FilterByEmailAddress(Predicate<string> filter) to return a list of names where the email address matches the filter. ✔

Use ISampleData.People for your data source. ✔
Implement ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people) to return a string that contains a unique, comma-separated list of states. ✔

Use the people parameter from ISampleData.People property for your data source. ✔
At a minimum, use the System.Linq.Enumerable.Aggregate` LINQ method to create your result. ✔
Don't forget the list should be unique. ✔
It is recommended that, at a minimum, you use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows to validate your result.
Given the implementation of Node in Assignment5

Implement IEnumerable<T> to return all the items in the "circle" of items. ✔
Add an IEnumberable<T> ChildItems(int maximum) method to Node that returns the remaining items with a maximum number of items returned less than maximum.
Extra Credit
Implement the homework using async/await and multi-threading by defining a new SampleDataAsync class that implements IAsyncSampleData). Refactor your SampleData and SampleDataAsync classes with minimal duplication. Be sure to refactor your tests to re-use a significant amount of the test code for both implementations. ❌
Fundamentals
Place all shared project properties into a Directory.Build.Props file.
Place all shared project items into a Directory.Build.targets file.
Ensure nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, consider using Assert.AreEqual<T>() (the generic version) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

File: Assignment.Tests/NodeTests.cs [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        //Clear the list and verify only start remains
+        node.Clear();
+        Assert.AreEqual("apple", node.Data);
+        Assert.AreNotEqual("pear", node.Next.Data);
+        Assert.AreNotEqual("orange", node.Next.Next.Data);
+
+        Assert.IsTrue(node.Exists("apple"));
+        Assert.IsFalse(node.Exists("pear"));
+        Assert.IsFalse(node.Exists("orange"));
+    }
```
Comment (omar5o9) [inline Assignment.Tests/NodeTests.cs:103]: Great test

**Review by omar5o9 — COMMENTED** [review]
Implement the ISampleData.CsvRows property, loading the data from the People.csv file and returning each line as a single string. ✔

Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔
Using LINQ, skip the first row in the People.csv. ✔
Be sure to appropriately handle resource (IDisposable) items correctly if applicable (and it may not be depending on how you implement it). ✔
Implement IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() to return a sorted, unique list of states. ✔

Use ISampleData.CsvRows for your data source. ✔
Don't forget the list should be unique. ✔
Sort the list alphabetically. ✔
Include a test that leverages a hardcoded list of addresses. ✔
Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ✔
Implement ISampleData.GetAggregateSortedListOfStatesUsingCsvRows() to return a string that contains a unique, comma separated list of states. ✔

Use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows() for your data source. ✔
Consider "selecting" only the states and calling ToArray() to retrieve an array of all the state names. ✔
Given the array, consider using string.Join to combine the list into a single string. ✔
Implement the ISampleData.People property to return all the items in People.csv as Person objects ✔

Use ISampleData.CsvRows as the source of the data. ✔
Sort the list by State, City, and Zip. (Sort the addresses first then select). ✔
Be sure that Person.Address is also populated. ✔
Adding null validation to all the Person and Address properties is optional.
Consider using ISampleData.CsvRows in your test to verify your results. ✔
Implement ISampleDate.FilterByEmailAddress(Predicate<string> filter) to return a list of names where the email address matches the filter. ✔

Use ISampleData.People for your data source. ✔
Implement ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people) to return a string that contains a unique, comma-separated list of states. ✔

Use the people parameter from ISampleData.People property for your data source. ✔
At a minimum, use the System.Linq.Enumerable.Aggregate` LINQ method to create your result. ✔
Don't forget the list should be unique. ✔
It is recommended that, at a minimum, you use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows to validate your result.
Given the implementation of Node in Assignment5

Implement IEnumerable<T> to return all the items in the "circle" of items. ✔
Add an IEnumberable<T> ChildItems(int maximum) method to Node that returns the remaining items with a maximum number of items returned less than maximum.

Fundamentals
Place all shared project properties into a Directory.Build.Props file.
Place all shared project items into a Directory.Build.targets file.
Ensure nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, consider using Assert.AreEqual<T>() (the generic version) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

Great job all around

File: Assignment.Tests/SampleDataTests.cs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    [TestMethod]
+    public void GetUniqueSortedListOfStatesGivenCsvRows_SortedAndUnique_ReturnsSuccessful()
+    {
+        List<string> uniqueSortedStates = SampleData.GetUniqueSortedListOfStatesGivenCsvRows().ToList();
+        Assert.AreEqual("AL", uniqueSortedStates[0]);
+        Assert.AreEqual("WV", uniqueSortedStates[26]);
+    }
+
```
Comment (Tfelchlin13) [inline Assignment.Tests/SampleDataTests.cs:32]: you need another test for GetUniqueSortedListOfStatesGivenCsvRows that uses hard code data this one uses non hard coded data for GetUniqueSortedListOfStatesGivenCsvRows.

File: Assignment.Tests/NodeTests.cs [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+    [TestMethod]
+    public void GetEnumerator_AddElements_ReturnsSuccessful()
+    {
+        Node<string> node = new("Hello there");
+        node.Append("General Kenobi");
+        node.Append("You are a bold one");
+
+        Assert.AreEqual(3, node.Count());
+        Assert.IsTrue(new List<string> { "Hello there", "General Kenobi", "You are a bold one"}.SequenceEqual(node.Select(Node => Node.Data)));
+    }
```
Comment (robertogarcia11) [inline Assignment.Tests/NodeTests.cs:121]: nice test!, "You Fool. I have been trained in your Jedi Arts..."

File: Assignment/SampleData.cs
Comment (robertogarcia11) [inline Assignment/SampleData.cs:1]: really nice indexing/tabing, makes the code look so clean and understandable.

File: Directory.Build.props (hunk 2 above)
Comment (robertogarcia11) [reply Directory.Build.props:4]: +1

**Comment by robertogarcia11** [comment]
Implement the ISampleData.CsvRows property, loading the data from the People.csv file and returning each line as a single string. ✔

Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔
Using LINQ, skip the first row in the People.csv. ✔
Be sure to appropriately handle resource (IDisposable) items correctly if applicable (and it may not be depending on how you implement it). ✔
Implement IEnumerable GetUniqueSortedListOfStatesGivenCsvRows() to return a sorted, unique list of states. ✔

Use ISampleData.CsvRows for your data source. ✔
Don't forget the list should be unique. ✔
Sort the list alphabetically. ✔
Include a test that leverages a hardcoded list of addresses. ✔
Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ✔
Implement ISampleData.GetAggregateSortedListOfStatesUsingCsvRows() to return a string that contains a unique, comma separated list of states. ✔

Use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows() for your data source. ✔
Consider "selecting" only the states and calling ToArray() to retrieve an array of all the state names. ✔
Given the array, consider using string.Join to combine the list into a single string. ✔
Implement the ISampleData.People property to return all the items in People.csv as Person objects ✔

Use ISampleData.CsvRows as the source of the data. ✔
Sort the list by State, City, and Zip. (Sort the addresses first then select). ✔
Be sure that Person.Address is also populated. ✔
Adding null validation to all the Person and Address properties is optional.
Consider using ISampleData.CsvRows in your test to verify your results. ✔
Implement ISampleDate.FilterByEmailAddress(Predicate filter) to return a list of names where the email address matches the filter. ✔

Use ISampleData.People for your data source. ✔
Implement ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable people) to return a string that contains a unique, comma-separated list of states. ✔

Use the people parameter from ISampleData.People property for your data source. ✔
At a minimum, use the System.Linq.Enumerable.Aggregate` LINQ method to create your result. ✔
Don't forget the list should be unique. ✔
It is recommended that, at a minimum, you use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows to validate your result.
Given the implementation of Node in Assignment5

Implement IEnumerable to return all the items in the "circle" of items. ✔
Add an IEnumberable ChildItems(int maximum) method to Node that returns the remaining items with a maximum number of items returned less than maximum.

Fundamentals
Place all shared project properties into a Directory.Build.Props file.
Place all shared project items into a Directory.Build.targets file.
Ensure nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, consider using Assert.AreEqual() (the generic version) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

Nice job!!

File: Assignment.Tests/SampleDataTests.cs (hunk 5 above)
Comment (Purplepigs0129) [reply Assignment.Tests/SampleDataTests.cs:32]: I totally missed this before I did the code review I will add one

File: Directory.Build.props (hunk 3 above)
Comment (Purplepigs0129) [reply Directory.Build.props:3]: Thanks

File: Directory.Build.props (hunk 2 above)
Comment (Purplepigs0129) [reply Directory.Build.props:4]: Thanks for the catch here

File: Assignment.Tests/Assignment.Tests.csproj [hunk 7]
```diff
@@ -0,0 +1,24 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net8.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
```
Comment (BenjaminMichaelis) [inline Assignment.Tests/Assignment.Tests.csproj:6]: Why are these here instead of just in the `Directory.Build.props` file?

File: Assignment.Tests/NodeTests.cs [hunk 8]
```diff
@@ -0,0 +1,139 @@
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+
+namespace Assignment.Test;
+
+#pragma warning disable CS8618
```
Comment (BenjaminMichaelis) [inline Assignment.Tests/NodeTests.cs:6]: Justification needed for disabling warnings

File: Assignment.Tests/NodeTests.cs [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+        Node<string> node = new("Hello there");
+        node.Append("General Kenobi");
+        node.Append("You are a bold one");
+
+        Assert.AreEqual(3, node.Count());
+        Assert.IsTrue(new List<string> { "Hello there", "General Kenobi", "You are a bold one"}.SequenceEqual(node.Select(Node => Node.Data)));
+    }
+
+    [TestMethod]
+    public void ChildItems_Input4SetMaxTo3_ReturnsSuccessful()
```
Comment (BenjaminMichaelis) [inline Assignment.Tests/NodeTests.cs:124]: What about edge cases? What if I give it a max of `-4`? Should it throw an exception, should it return empty?

File: Assignment/Node.cs [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+    public IEnumerable<Node<T>> ChildItems(int maximum)
+    {
+        int count = 0;
+        foreach(var value in this.Skip(1))
+        {
+            if (count++ == maximum) break;
+            yield return value;
+        }
```
Comment (BenjaminMichaelis) [inline Assignment/Node.cs:75]: I think this could be simplified just using linq

File: Assignment/SampleData.cs [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+
+    // 3.
+    public string GetAggregateSortedListOfStatesUsingCsvRows() => string
+        .Join(", ", GetUniqueSortedListOfStatesGivenCsvRows().ToArray());
+
+    // 4.
+    public IEnumerable<IPerson> People => CsvRows
+        .Select(line => line.Split(","))
+        .OrderBy(line => (line[5], line[6], line[7]))
+        .Select(line => new Person(line[1], line[2], new Address(line[4], line[5], line[6], line[7]), line[3]));
```
Comment (BenjaminMichaelis) [inline Assignment/SampleData.cs:28]: for specific columns in this file, a class or enum would be nice to give readable values instead of like '6' or '2' to different columns, so you could say like `[Column.State]` or `[Column.City]`, etc.
